content/examples/01-next-app/06-assistants/02-stream-assistant-responses-with-tools.mdx

```diff
@@ -0,0 +1,278 @@
+      ...currentMessages,
+      {
+        id: '123',
+        status: 'user.message.created',
+        text: input,
+        gui: null,
+      },
+    ]);
+
+    const response = await submitMessage(input);
+    setMessages(currentMessages => [...currentMessages, response]);
+    setInput('');
+  };
+
+  return (
+    <div className="flex flex-col-reverse">
+      <div className="flex flex-row gap-2 p-2 bg-zinc-100 w-full">
+        <input
+          className="bg-zinc-100 w-full p-2 outline-none"
+          value={input}
+          onChange={event => setInput(event.target.value)}
+          placeholder="Ask a question"
+          onKeyDown={event => {
+            if (event.key === 'Enter') {
+              handleSubmission();
+            }
+          }}
+        />
+        <button
+          className="p-2 bg-zinc-900 text-zinc-100 rounded-md"
+          onClick={handleSubmission}
+        >
+          Send
+        </button>
+      </div>
+
+      <div className="flex flex-col h-[calc(100dvh-56px)] overflow-y-scroll">
+        <div>
+          {messages.map(message => (
+            <div key={message.id} className="flex flex-col gap-1 border-b p-2">
+              <div className="flex flex-row justify-between">
+                <div className="text-sm text-zinc-500">{message.status}</div>
+              </div>
+              <div className="flex flex-col gap-2">{message.gui}</div>
+              <div>{message.text}</div>
+            </div>
+          ))}
+        </div>
+      </div>
+    </div>
+  );
+}
+```
+
+```tsx filename='app/message.tsx'
+'use client';
+
+import { StreamableValue, useStreamableValue } from 'ai/rsc';
+
+export function Message({ textStream }: { textStream: StreamableValue }) {
+  const [text] = useStreamableValue(textStream);
+
+  return <div>{text}</div>;
+}
+```
+
+```tsx filename='app/message.tsx'
+'use client';
+
+import { StreamableValue, useStreamableValue } from 'ai/rsc';
+
+export function Message({ textStream }: { textStream: StreamableValue }) {
+  const [text] = useStreamableValue(textStream);
+
+  return <div>{text}</div>;
+}
+```
+
+## Server
+
+In your server action, you will create a function called `submitMessage` that add the user's message to the thread. The function will create a new thread if one does not exist and add the user's message to the thread. If a thread already exists, the function will add the user's message to the existing thread. The function will then create a run and stream the assistant's response to the client. Furthermore, the run queue is used to manage multiple runs in the same thread during the lifetime of the server action.
+
+In case the assistant requires a tool call, the server action will handle the tool call and return the output to the assistant. In this example, the assistant requires a tool call to search for emails. The server action will search for emails based on the query and has_attachments parameters and return the output to the assistant.
```

Comment:
```suggestion
In case the assistant requires a tool call, the server action will handle the tool call and return the output to the assistant. In this example, the assistant requires a tool call to search for emails. The server action will search for emails based on the query and `has_attachments` parameters and return the output to the assistant.
```